File: LLD/LLD-2/src/TicTakToeGame/src/main/java/org/example/models/Game.java
package org.example.models;

import org.example.exceptions.GameInvalidException;
import org.example.strategies.winningStrategies.WinningStrategy;

import java.util.ArrayList;
import java.util.List;

public class Game {
    private Board board;
    List<Player> players;

    private GameStatus gameStatus;

    private Player winner;

    private int nextMovePlayerIndex;  //which player is having next turn

    private List<Move> moves;

    private List<WinningStrategy> winningStrategies;

    //private because of builder dp
    //Builder --> HW

    public static Builder getBuilder()
    {
        return new Builder();
    }
    private Game(int dimension, List<Player> players, List<WinningStrategy> winningStrategies) {
        this.board = new Board(dimension);
        this.gameStatus = GameStatus.IN_PROGRESS;
        this.nextMovePlayerIndex =0;
        this.moves = new ArrayList<>();
        this.winningStrategies = winningStrategies;
        this.players = players;
    }

    //Builder inner class
    public static class Builder
    {
        private int dimension;
        private List<Player> players;
        private List<WinningStrategy> winningStrategies;

        private Builder()
        {
            this.players = new ArrayList<>();
            this.winningStrategies = new ArrayList<>();
            this.dimension =0;
        }

        public Builder setDimension(int dimension) {
            this.dimension = dimension;
            return this;
        }

        public Builder setPlayers(List<Player> players) {
            this.players = players;
            return this;
        }

        public Builder setWinningStrategies(List<WinningStrategy> winningStrategies) {
            this.winningStrategies = winningStrategies;
            return this;
        }

        private static boolean validate()
        {
            //Validations
            //1. Only one Bot is allowed per game.
            //2. NO 2 players should have the same symbol
            return true;
        }
        public Game build() throws Exception {
            if(!validate())
                throw new Exception();
            return new Game(this.dimension, this.players, this.winningStrategies);
        }
    }

    public Board getBoard() {
        return board;
    }

    public void setBoard(Board board) {
        this.board = board;
    }

    public List<Player> getPlayers() {
        return players;
    }

    public void setPlayers(List<Player> players) {
        this.players = players;
    }

    public GameStatus getGameStatus() {
        return gameStatus;
    }

    public void setGameStatus(GameStatus gameStatus) {
        this.gameStatus = gameStatus;
    }

    public Player getWinner() {
        return winner;
    }

    public void setWinner(Player winner) {
        this.winner = winner;
    }

    public int getNextMovePlayerIndex() {
        return nextMovePlayerIndex;
    }

    public void setNextMovePlayerIndex(int nextMovePlayerIndex) {
        this.nextMovePlayerIndex = nextMovePlayerIndex;
    }

    public List<Move> getMoves() {
        return moves;
    }

    public void setMoves(List<Move> moves) {
        this.moves = moves;
    }

    public void printBoard()
    {
        board.printBoard();
    }

    public void undo()
    {

    }
    private boolean validateMove(Move move)
    {
        int row = move.getCell().getRow();
        int col = move.getCell().getCol();

/*
        if(row >= board.getSize() || row <0 || col >= getBoard().getSize() || col < 0
        || !board.getBoard().get(row).get(col).getCellState().equals(CellState.EMPTY))
        {
            return false;
        }
        return true;
*/

        return row < board.getSize() && row >= 0 && col < getBoard().getSize() && col >= 0
                && board.getBoard().get(row).get(col).getCellState().equals(CellState.EMPTY);
    }
    public void makeMove(Board board)
    {
        //Current Player to make the move
        Player currentPlayer = players.get(nextMovePlayerIndex);

        System.out.println("It is "+ currentPlayer.getName()+"'s move");

        Move move = currentPlayer.makeMove(this.board);

        System.out.println(currentPlayer.getName() + "has a made a move at Row: "+ move.getCell().getRow()
        + "and at Column: "+ move.getCell().getCol()+" .");

        //Validate the move before we apply the move on board.
        if(!validateMove(move))
        {
            System.out.println("Invalid move by player: " + currentPlayer.getName());
            return;
        }
        int row = move.getCell().getRow();
        int col = move.getCell().getCol();

        Cell finalCell = board.getBoard().get(row).get(col);
        finalCell.setCellState(CellState.FILLED);
        finalCell.setPlayer(currentPlayer);

        Move finalMove = new Move(finalCell, currentPlayer); //valid move some moves might not be valid
        moves.add(move);

        nextMovePlayerIndex +=1;
        nextMovePlayerIndex %= players.size();

        //Once player has made a move, check the winner
        //checkWinner

        if(checkWinner(board, finalMove))
        {
            gameStatus = GameStatus.ENDED;
            winner = currentPlayer;
        } else if(moves.size() == board.getSize() * board.getSize())
        {
            gameStatus = GameStatus.DRAW;
        }
    }
    private boolean checkWinner(Board board, Move move)
    {
        for(WinningStrategy winningStrategy : winningStrategies)
        {
            if(winningStrategy.checkWinner(move , board))
            {
                return true;
            }
        }
        return false;
    }
}

File: LLD3/TicTacToe/TicTacToe/Models/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Remoting.Messaging;
using System.Text;
using System.Threading.Tasks;
using TicTacToe.Strategies.WinningStrategies;

namespace TicTacToe.Models
{
    public class Game
    {
        private List<Player> _players;

        private Board _board;

        private List<Move> _moves;

        private Player _winner;

        private GameStatus _gameStatus;

        private List<IWinningStrategy> _winningStrategies;

        private int _nextMovePlayerIndex;
        public List<Player> Players { get => _players; set => _players = value; }
        public Board Board { get => _board; set => _board = value; }
        public Player Winner { get => _winner; set => _winner = value; }
        public List<IWinningStrategy> WinningStrategies { get => _winningStrategies; set => _winningStrategies = value; }
        public GameStatus GameStatus { get => _gameStatus; set => _gameStatus = value; }
        public int NextMovePlayerIndex { get => _nextMovePlayerIndex; set => _nextMovePlayerIndex = value; }

        public static Builder GetBuilder()
        {
            return new Builder();
        }

        public class Builder
        {
            private int _dimension;

            private List<Player> _players;

            private List<IWinningStrategy> _winningStrategies;

            public Builder SetDimension(int dimension)
            {
                this._dimension = dimension;
                return this;
            }

            public Builder SetPlayers(List<Player> players)
            {
                this._players = players;
                return this;
            }

            public Builder SetWinningStretegies(List<IWinningStrategy> winningStrategies)
            {
                this._winningStrategies = winningStrategies;
                return this;
            }

            private bool Validate()
            {
                return true;
            }

            public Game Build()
            {
                if (!Validate())
                    throw new Exception("You are Idiot"); //TODO
                return new Game(_dimension, _players, _winningStrategies);
            }
        }

        private Game(int dimension, List<Player> players, List<IWinningStrategy> winningStrategies)
        {
            this._board = new Board(dimension);
            this._gameStatus = GameStatus.In_Progress;
            this._nextMovePlayerIndex = 0;

            this._moves = new List<Move>();

            this._winningStrategies = winningStrategies;
            this._players = players;
        }

        public void DisplayBoard()
        {
            _board.DisplayBoard();
        }

        public void Undo()
        {

        }

        private bool ValidateMove(Move move)
        {

            int row = move.Cell.Row;
            int col = move.Cell.Col;

            return row < _board.Dimension && row >= 0 && col < Board.Dimension && col >= 0 && _board.Boards[row, col].State.Equals(CellState.Empty);
        }
        private bool CheckWinner(Board board, Move move)
        {
            foreach (IWinningStrategy winningStrategy in _winningStrategies)
            {
                if (winningStrategy.CheckForWinner(move, board))
                    return true;
            }
            return false;
        }

        public void MakeMove()
        {
            Player currentPlayer = _players[_nextMovePlayerIndex];

            Console.WriteLine("It is " + currentPlayer.Name + "'s move");
            
            Move move = currentPlayer.MakeMove(this._board);

            Console.WriteLine(currentPlayer.Name + " has made a move at index row = " + move.Cell.Row + ", col = " + move.Cell.Col + ".");

            if (!ValidateMove(move))
            {
                Console.WriteLine("Invalid move by player: " + currentPlayer.Name);
            }

            int row = move.Cell.Row;
            int col = move.Cell.Col;

            Cell finalCellToMakeMove = _board.Boards[row, col];
            finalCellToMakeMove.State = CellState.Filled;
            finalCellToMakeMove.Player = currentPlayer;

            Move finalMove = new Move(currentPlayer, finalCellToMakeMove);
            _moves.Add(finalMove);

            _nextMovePlayerIndex += 1;
            _nextMovePlayerIndex %= _players.Count;

            if (CheckWinner(_board, finalMove))
            {
                _gameStatus = GameStatus.Ended;
                _winner = currentPlayer;
            }
            else if (_moves.Count == _board.Dimension * _board.Dimension)
            {
                _gameStatus = GameStatus.Draw;
            }
        }
    }
}
